### PR TITLE
Create Pull Policy on Installer Image During Cluster Provisioning

### PR DESCRIPTION
**Pull request by bennerv**
Story isn't groomed yet.  Related card / ask from ARO team: https://issues.redhat.com/browse/HIVE-2209

### CODE REVIEW

File: pkg/install/generate.go [hunk 1]
```diff
@@ -325,7 +325,7 @@ func InstallerPodSpec(
 		{
 			Name:            "cli",
 			Image:           cliImage,
-			ImagePullPolicy: corev1.PullIfNotPresent,
+			ImagePullPolicy: images.GetInstallerClusterProvisionImagePullPolicy(),
```
Comment (2uasimojo) [inline pkg/install/generate.go:328]: I think you meant for this to go on L317.

File: pkg/controller/images/controller_images.go [hunk 2]
```diff
@@ -47,6 +53,13 @@ func GetHiveClusterProvisionImagePullPolicy() corev1.PullPolicy {
 	return corev1.PullPolicy(envVarOrDefault(HiveClusterProvisionImagePullPolicyEnvVar, string(corev1.PullIfNotPresent)))
 }
 
+// GetInstallerClusterProvisionImagePullPolicy returns the policy to use when pulling the installer image
+// for the ClusterProvision pod.
+// Either the one specified in the environment variable or the hardcoded default.
+func GetInstallerClusterProvisionImagePullPolicy() corev1.PullPolicy {
+	return corev1.PullPolicy(envVarOrDefault(InstallerClusterProvisionImagePullPolicyEnvVar, string(corev1.PullIfNotPresent)))
```
Comment (2uasimojo) [inline pkg/controller/images/controller_images.go:60]: We would also need a way to get this env var down to the controller. In fact, it looks to me like the `HiveClusterProvisionImagePullPolicyEnvVar` (upon which I'm guessing you based this change) can't be used today for that same reason.

**Review by 2uasimojo — COMMENTED** [review]
FWIW, when I glanced at this I was considering adding an `InstallerImagePullPolicy` field right next to the existing [InstallerImageOverride](https://github.com/openshift/hive/blob/7cbb91212d8a79ad33c87d89d823d84ae3983f68/apis/hive/v1/clusterdeployment_types.go#L226-L229). The installmanager pod already has the ClusterDeployment, so that would be easy to read, vs having to percolate an env var down through three layers of invocation (hive-operator => hive-controllers => installer pod).

**Comment by 2uasimojo** [comment]
Hijacking this PR to sniff out some CI issues...

/test e2e-azure
/test e2e-gcp

File: pkg/imageset/updateinstaller.go [hunk 3]
```diff
@@ -182,6 +182,12 @@ func (o *UpdateInstallerImageOptions) Run() (returnErr error) {
 		o.log.WithField("installerImage", installerImage).Info("installer image found")
 	}
 
+	installerImagePullPolicy := string(corev1.PullIfNotPresent)
```
Comment (bennerv) [inline pkg/imageset/updateinstaller.go:185]: This is where the default is.  I'm not sure if I should pull it up into a const to make it more visible.

File: pkg/imageset/updateinstaller.go [hunk 4]
```diff
@@ -182,6 +182,12 @@ func (o *UpdateInstallerImageOptions) Run() (returnErr error) {
...
+	if cd.Spec.Provisioning != nil && cd.Spec.Provisioning.InstallerImagePullPolicyOverride != "" {
+		installerImagePullPolicy = cd.Spec.Provisioning.InstallerImagePullPolicyOverride
+		o.log.WithField("installer image pull policy", installerImagePullPolicy).Info("installer image pull policy overridden")
```
Comment (2uasimojo) [inline pkg/imageset/updateinstaller.go:188]: nit: I think keys for log fields are usually made without spaces so as to make the log line more easily parseable/greppable.

File: pkg/imageset/updateinstaller.go (hunk 3 above)
Comment (2uasimojo) [reply pkg/imageset/updateinstaller.go:185]: Shrug, if ya like, I'm fine either way.

File: pkg/imageset/updateinstaller.go [hunk 5]
```diff
@@ -221,6 +227,7 @@ func (o *UpdateInstallerImageOptions) Run() (returnErr error) {
 	cd.Status.InstallerImage = &installerImage
 	cd.Status.CLIImage = &cliImage
 	cd.Status.InstallVersion = &releaseVersion
+	cd.Status.InstallerImagePullPolicy = &installerImagePullPolicy
```
Comment (2uasimojo) [inline pkg/imageset/updateinstaller.go:230]: femtonit: might move this up next to InstallerImage @L227.

**Review by 2uasimojo — COMMENTED** [review]
This looks good. Couple of nonblocking items inline.

However, please see recent comment on the [card](https://issues.redhat.com/browse/HIVE-2209): this would not be necessary if InstallerImageOverride were specified by digest rather than by tag. Is there a good reason that's not done?

**Comment by bennerv** [comment]
@2uasimojo 
>this would not be necessary if InstallerImageOverride were specified by digest rather than by tag. Is there a good reason that's not done?

I had a comment typed up last night around this.  

I'm going to see ARO's level of effort to implement this on ARO's side for production vs. this PR change.  We're going to shift to the full digest eventually, just need to confirm if we need this as a stop gap temporarily or not.

**Comment by 2uasimojo** [comment]
Cool. If it's just a stopgap, then I would greatly prefer it not be made part of the official API, cause we can't undo that once it's in place. It could instead be an annotation on the CD. And in that case I think we should be able to do without the status field as well, right?

**Comment by bennerv** [comment]
@2uasimojo I'm going to close this as I think we can do it from our side with a similar amount of effort and it saves some code changes from hive's side as well.

**Pull request closed without merging**